### PR TITLE
[WIP] Add bfloat16 support #243

### PR DESCRIPTION
**Pull request by ikergarcia1996**
Implement bfloat16 support (#243). I added a new flag named "mixed precision" that can be "no" for FP32 training. "fp16" for FP16 training and "bf16" for bfloat16 training. This version of the code is compatible with previous scripts/config files using accelerate, the fp16 flag has the same behavior as before (Should the users be advice to use the new flag when they use the fp16 flag with a deprecation warning?)

I tested the code using a T5-large model. FP16 mixed precision results NaN loss, thanks to bfloat16 training using mixed precision is possible. 

image

### CODE REVIEW

File: src/accelerate/utils.py [hunk 1]
```diff
@@ -240,23 +240,23 @@ def _initialize_tensor(tensor_info):
...
+        The same data structure as :obj:`tensor` with all tensors that were in FP16/BF16 precision converted to FP32.
     """
 
     def _convert_to_fp32(tensor):
         return tensor.float()
 
-    def _is_fp16_tensor(tensor):
-        return hasattr(tensor, "dtype") and tensor.dtype == torch.float16
+    def _is_fp16_bf16_tensor(tensor):
+        return hasattr(tensor, "dtype") and (tensor.dtype == torch.float16 or tensor.dtype == torch.bfloat16)
```
Comment (sgugger) [inline src/accelerate/utils.py:257]: This would also need a version check, as I'm not sure the `torch.bfloat16` was there in PyTorch 1.6.

**Review by sgugger — APPROVED** [review]
Thanks a lot for your PR, this is great work!

I'm okay with the deprecation cycle, which, coupled with an automatic update of the configs should be pretty smooth for the user.

A few important fixes to bring in the PR:
- make sure to only use the env variable when the `mixed_precision` attribute is not set (we shouldn't overrride a `"no"` passed by the user), and that this attributes default to `None`.
- make sure the bfloat16 functionality is only accessed when we have PyTorch 1.10 or later
I've left comments for those.

File: src/accelerate/accelerator.py [hunk 2]
```diff
@@ -311,7 +348,12 @@ def prepare_model(self, model):
             kwargs = self.ddp_handler.to_kwargs() if self.ddp_handler is not None else {}
             model = torch.nn.parallel.DistributedDataParallel(model, **kwargs)
         if self.native_amp:
-            model.forward = torch.cuda.amp.autocast()(model.forward)
+            if self.mixed_precision == "fp16":
+                model.forward = torch.cuda.amp.autocast(dtype=torch.float16)(model.forward)
```
[the remark this answers is not in the document]
Comment (ikergarcia1996) [reply src/accelerate/accelerator.py:354]: Version test added

File: src/accelerate/commands/config/config_args.py [hunk 3]
```diff
@@ -97,6 +98,12 @@ def from_yaml_file(cls, yaml_file=None):
             config_dict = yaml.safe_load(f)
         if "compute_environment" not in config_dict:
             config_dict["compute_environment"] = ComputeEnvironment.LOCAL_MACHINE
+        if "mixed_precision" not in config_dict:
```
[the remark this answers is not in the document]
Comment (ikergarcia1996) [reply src/accelerate/commands/config/config_args.py:105]: I forgot about the "from_json_file" method, i have updated it too and I added a line to delete the fp16 key from old config files

**Comment by ikergarcia1996** [comment]
I have updated the PR with the fixes you mention

File: src/accelerate/accelerator.py [hunk 4]
```diff
@@ -101,12 +106,29 @@ def __init__(
...
         fp16: bool = None,
+        mixed_precision: str = None,
         cpu: bool = False,
         deepspeed_plugin: DeepSpeedPlugin = None,
         rng_types: Optional[List[Union[str, RNGType]]] = None,
         dispatch_batches: Optional[bool] = None,
         kwargs_handlers: Optional[List[KwargsHandler]] = None,
     ):
+
+        if mixed_precision:
```
Comment (sgugger) [inline src/accelerate/accelerator.py:117]:
```suggestion
        if mixed_precision is not None:
```
We usually prefer explicit tests instead of relying on Python's bool conversion.

File: src/accelerate/accelerator.py [hunk 5]
```diff
@@ -101,12 +106,29 @@ def __init__(
...
+            mixed_precision = mixed_precision.lower()
+            if mixed_precision not in [
+                "no",
+                "fp16",
+                "bf16",
+            ]:
```
Comment (sgugger) [inline src/accelerate/accelerator.py:123]:
```suggestion
            if mixed_precision not in ["no", "fp16", "bf16"]:
```
Should fit in one line now.

File: src/accelerate/accelerator.py [hunk 6]
```diff
@@ -149,10 +175,14 @@ def __init__(
...
         self.native_amp = False
-        if self.state.use_fp16:
+        if self.state.mixed_precision == "fp16":
             self.native_amp = version.parse(torch.__version__) >= version.parse("1.6")
             kwargs = self.scaler_handler.to_kwargs() if self.scaler_handler is not None else {}
             self.scaler = torch.cuda.amp.GradScaler(**kwargs)
+        elif self.state.mixed_precision == "bf16":
+            self.native_amp = version.parse(torch.__version__) >= version.parse("1.10")
+            kwargs = self.scaler_handler.to_kwargs() if self.scaler_handler is not None else {}
+            self.scaler = torch.cuda.amp.GradScaler(**kwargs)
```
Comment (sgugger) [inline src/accelerate/accelerator.py:185]: I realize we are silently not training in mixed precision when:
- `mixed_precision="fp16"` and torch version is < 1.6
- `mixed_precision="bf16"` and torch version is < 1.10

I think we should raise an error message in both cases, or at least a warning.

File: src/accelerate/commands/config/config_args.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sgugger) [reply src/accelerate/commands/config/config_args.py:105]: Nice!

File: src/accelerate/commands/launch.py [hunk 7]
```diff
@@ -53,9 +54,20 @@ def launch_command_parser(subparsers=None):
...
+        choices=["no", "fp16", "bf16"],
+        help="Whether or not to use mixed precision training. "
+        "Choose between FP16 and BF16 (bfloat16) training. "
+        "BF16 training is only supported on Nvidia Ampere GPUs and PyTorch 1.10 or later.",
+    )
+
     parser.add_argument(
         "--fp16", default=False, action="store_true", help="Whether or not to use mixed precision training."
     )
+
```
Comment (sgugger) [inline src/accelerate/commands/launch.py:70]: No need for this new line I think.
```suggestion
```

File: src/accelerate/commands/launch.py [hunk 8]
```diff
@@ -134,7 +146,20 @@ def simple_launcher(args):
...
     current_env = os.environ.copy()
     current_env["USE_CPU"] = str(args.cpu)
-    current_env["USE_FP16"] = str(args.fp16)
+
+    mixed_precision = args.mixed_precision.lower()
+    if mixed_precision not in [
+        "no",
+        "fp16",
+        "bf16",
+    ]:
```
Comment (sgugger) [inline src/accelerate/commands/launch.py:155]:
```suggestion
    if mixed_precision not in ["no", "fp16", "bf16"]:
```

File: src/accelerate/commands/launch.py [hunk 9]
```diff
@@ -168,7 +193,20 @@ def multi_gpu_launcher(args):
...
 
     current_env = os.environ.copy()
-    current_env["USE_FP16"] = str(args.fp16)
+    mixed_precision = args.mixed_precision.lower()
+
+    if mixed_precision not in [
+        "no",
+        "fp16",
+        "bf16",
+    ]:
```
Comment (sgugger) [inline src/accelerate/commands/launch.py:202]:
```suggestion
    if mixed_precision not in ["no", "fp16", "bf16"]:
```

File: src/accelerate/commands/launch.py [hunk 10]
```diff
@@ -201,7 +239,20 @@ def deepspeed_launcher(args):
...
```
Comment (sgugger) [inline src/accelerate/commands/launch.py:248]:
```suggestion
    if mixed_precision not in ["no", "fp16", "bf16"]:
```

File: src/accelerate/commands/launch.py [hunk 11]
```diff
@@ -309,8 +360,21 @@ def sagemaker_launcher(sagemaker_config: SageMakerConfig, args):
...
     hyperparameters = _convert_nargs_to_dict(args.training_script_args)
 
-    environment = {"USE_FP16": args.fp16}  # Environment variables to be set for use during training job
+    mixed_precision = args.mixed_precision.lower()
+
+    if mixed_precision not in [
+        "no",
+        "fp16",
+        "bf16",
+    ]:
```
Comment (sgugger) [inline src/accelerate/commands/launch.py:369]:
```suggestion
    if mixed_precision not in ["no", "fp16", "bf16"]:
```

File: src/accelerate/notebook_launcher.py [hunk 12]
```diff
@@ -39,6 +40,8 @@ def notebook_launcher(function, args=(), num_processes=None, use_fp16=False, use
             the number of GPUs available otherwise.
         use_fp16 (:obj:`bool`, `optional`, defaults to :obj:`False`):
             If :obj:`True`, will use mixed precision training on multi-GPU.
```
Comment (sgugger) [inline src/accelerate/notebook_launcher.py:42]: You can remove this from the documentation (we try to avoid documenting deprecated args so users don't start using them).

File: src/accelerate/notebook_launcher.py [hunk 13]
```diff
@@ -110,7 +113,22 @@ def notebook_launcher(function, args=(), num_processes=None, use_fp16=False, use
...
             os.environ["MASTER_ADDR"] = "127.0.0.1"
             os.environ["MASTER_PORT"] = str(use_port)
-            os.environ["USE_FP16"] = str(use_fp16)
+
+            mixed_precision = mixed_precision.lower()
+            if mixed_precision not in [
+                "no",
+                "fp16",
+                "bf16",
+            ]:
```
Comment (sgugger) [inline src/accelerate/notebook_launcher.py:122]:
```suggestion
            if mixed_precision not in ["no", "fp16", "bf16"]:
```

**Review by sgugger — APPROVED** [review]
Thanks for iterating, I have a few more comments, mostly regarding styling.

**Comment by ikergarcia1996** [comment]
I think they are solved now. Maybe we should add a warning in 'cluster.py' to warn if the user chooses the "fp16" or "bf16" option but doesn't have the correct version of Pytorch installed?

**Comment by sgugger** [comment]
This [comment](https://github.com/huggingface/accelerate/pull/247/files#r800669214) and this [one](https://github.com/huggingface/accelerate/pull/247/files#r800671521) don't seem to have been addressed. Could you take care of them?
Quality check is also still failing.

We can also add warnings when the options are chosen, yes!

**Comment by ikergarcia1996** [comment]
Sorry the previous commit to solve the problem of silently not training in mixed precision was incorrect. I have a question about implementation. If  the version is correct we set self.native_amp to True. 
```Python
self.native_amp = version.parse(torch.__version__) >= version.parse("1.6")
```
If the version is incorrect we disable mixed precision by setting that variable to False. Is it used for anything else? If we are going to raise an error, should we remove it? 
Should we raise a error always if the pytorch version is not correct? I'm not sure I understand this part of the code 100% and I don't know if I've implemented right the errors if the pytorch version is not compatible with the chosen mixed precision.

**Comment by sgugger** [comment]
The way you did it is great! Thanks again for your awesome contribution :-)
